ccip-sdk/src/ton/exec.ts

```diff
@@ -0,0 +1,98 @@
+import { Address, beginCell, toNano } from '@ton/core'
+import { type TonClient, internal } from '@ton/ton'
+
+import type { ExecutionReport } from '../types.ts'
+import {
+  type CCIPMessage_V1_6_TON,
+  type TONWallet,
+  MANUALLY_EXECUTE_OPCODE,
+  serializeExecutionReport,
+} from './types.ts'
+import { waitForTransaction } from './utils.ts'
+
+/**
+ * Generates an unsigned execute report payload for the TON OffRamp contract.
+ *
+ * @param offRamp - OffRamp contract address.
+ * @param execReport - Execution report containing the CCIP message and proofs.
+ * @param opts - Optional execution options. Gas limit override for execution (0 = no override).
+ * @returns Object with target address, value, and payload cell.
+ */
+export function generateUnsignedExecuteReport(
+  offRamp: string,
+  execReport: ExecutionReport<CCIPMessage_V1_6_TON>,
+  opts?: { gasLimit?: number },
+): {
+  to: string
+  value: bigint
+  body: ReturnType<typeof beginCell>['endCell'] extends () => infer R ? R : never
+} {
+  // Serialize the execution report
+  const serializedReport = serializeExecutionReport(execReport)
+
+  // Use provided gasLimit as override, or 0 for no override
+  const gasOverride = opts?.gasLimit ? BigInt(opts.gasLimit) : 0n
+
+  // Construct the OffRamp_ManuallyExecute message
+  const payload = beginCell()
+    .storeUint(MANUALLY_EXECUTE_OPCODE, 32) // Opcode for OffRamp_ManuallyExecute
+    .storeUint(0, 64) // queryID (default 0)
+    .storeRef(serializedReport) // ExecutionReport as reference
+    .storeCoins(gasOverride) // gasOverride (optional, 0 = no override)
+    .endCell()
+
+  return {
+    to: offRamp,
+    value: toNano('0.5'),
+    body: payload,
+  }
+}
+
+/**
+ * Executes a CCIP message on the TON OffRamp contract.
+ * Serializes the execution report, constructs the OffRamp_ManuallyExecute message,
+ * sends the transaction via the wallet, and waits for confirmation.
+ *
+ * @param client - TonClient instance for RPC calls.
+ * @param wallet - TON wallet with contract and keypair for signing.
+ * @param offRamp - OffRamp contract address.
+ * @param execReport - Execution report containing the CCIP message and proofs.
+ * @param opts - Optional execution options. Gas limit override for execution (0 = no override).
+ * @returns Transaction hash in format "workchain:address:lt:hash".
+ */
+export async function executeReport(
+  client: TonClient,
+  wallet: TONWallet,
+  offRamp: string,
+  execReport: ExecutionReport<CCIPMessage_V1_6_TON>,
+  opts?: { gasLimit?: number },
+): Promise<{ hash: string }> {
+  const unsigned = generateUnsignedExecuteReport(offRamp, execReport, opts)
```

Comment:
Nit, just for future reference/style/arch alignment:
ideally, you can implement these "send-only" calls inside the class itself, and using `this.generateUnsignedExecuteReport` method directly, JUST IN CASE someone decides to subclass `TONChain` and implement a slightly different `generateUnsignedExecuteReport` (matching interface), and get it used in the `executeReport`, as it'd be expected from proper subclassing;

If you call the function directly (as you done here), then they'll need to override the 2nd method as well, to make sure it calls their 1st.

I don't think you need to change this now, but just so we can get aligned moving forward on these A->B split methods backed by standalone functions (when possible), like this